compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -398,133 +398,65 @@ impl<'tcx> Const<'tcx> {
         }
     }
 
-    /// Returns the evaluated constant
-    #[inline]
-    pub fn eval(
-        self,
-        tcx: TyCtxt<'tcx>,
-        param_env: ParamEnv<'tcx>,
-        span: Span,
-    ) -> Result<(Ty<'tcx>, ValTree<'tcx>), ErrorHandled> {
-        self.eval_valtree(tcx, param_env, span).map_err(|err| {
-            match err {
-                Either::Right(err) => err,
-                Either::Left(_bad_ty) => {
-                    // This can happen when we run on ill-typed code.
-                    let e = tcx.dcx().span_delayed_bug(
-                        span,
-                        "`ty::Const::eval` called on a non-valtree-compatible type",
-                    );
-                    e.into()
-                }
-            }
-        })
-    }
-
     /// Normalizes the constant to a value or an error if possible.
     #[inline]
-    pub fn normalize(self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Self {
-        match self.eval(tcx, param_env, DUMMY_SP) {
+    pub fn normalize_internal(self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Self {
```

Comment:
FWIW part of the reason for having `eval` and `normalize` was to match the API on `mir::Const`.

If there are good reasons to equip the two `Const` types with different API, we should do it. But I wonder if there's some good place to write down the rationale for that?

Reply:
This issue (and related ones) should probably be linking to: #130704

with `min_generic_const_args` and `associated_const_equality` featuers normalization of type system constants will be behaving much more like types. They'll return nested goals, access in scope whjere clauses such as `T: Trait<ASSOC = 10>` in order to normalize instead of simply "evaluating".

This means that the only correct way to normalize a `ty::Const` will be to use the "normal" normalization routines such as `normalize_erasing_regions` or `infcx/ocx/fcx.normalize`. With that in mind all of the `eval_x` and `normalize` methods on `ty::Const` become massive footguns as they are never correct to use.

Reply:
It would probably make sense to write this rationale down somewhere *in rustc* but I'm not sure where a good place for those docs to live would be

Reply:
Ah, I see -- yeah I started using `normalize` here since these functions do put the constant into a "normal form" of sorts, but this was not intended to match the type system normalization. I can see how that is confusing.

`eval` is not existing type system terminology though?

Maybe `mir::Const::normalize` should be renamed to something else to avoid this terminology clash. I just can't think of a better name...

Reply:
Fundamentally, I think what happens here is that we have two forms of normalization here (type system normalization reducing e.g. associated const paths, and const eval normalization evaluating a fully "type system normalized" const path to its normal form). The type system grabbed the term "normalize" as if it was the only kid in town, so now we have to figure out what terminology other systems can use.

Reply:
fwiw its not *just* a naming issue- if we were to try and update the `eval_x` methods to be the kind of normalization that the type system requires then we would have to return the nested goals of the normalization to the caller which in the past has caused soundness bugs.

That is the most significant motivation for this refactor- it makes it much more likely that const generics doesn't have subtle bugs around normalization of const aliases. Though I do also think having `ty::Const` have symmetry with `ty::Ty` is much nicer than having symmetry with `mir::Const` as it makes working with consts similar to working with types and generally I've found that where there's divergence in logic between type system consts and types, there tends to be bugs.

Reply:
I'm not really sure what a better name for `mir::Const::normalize` would be, it honestly might be fine in its current name since it *does* result in a fully normalized form since in codegen the only thing normalization has to do is evaluate the constant

Reply:
> Though I do also think having ty::Const have symmetry with ty::Ty is much nicer than having symmetry with mir::Const

As someone working on ty::Const and ty::Ty you would say that. As someone working on mir::Const and ty::Const, I disagree. ;) There'll also be bugs if mir::Const and ty::Const are too different and people assume they work the same.

Reply:
> I'm not really sure what a better name for mir::Const::normalize would be, it honestly might be fine in its current name since it does result in a fully normalized form since in codegen the only thing normalization has to do is evaluate the constant

It's not just used in codegen, though. AFAIK it's also used in various places during the MIR pipeline where things are generic, but we'd like to have the const value evaluated if possible, and only store the `Unevaluated` const identity if truly needed.

Maybe `eval_if_possible` would be a better name? However, you seem to say that the `eval` methods on `ty::Const` are in general problematic, which is quite concerning since `mir::Const::eval` needs to call `ty::Const::eval`, and if we can't evaluate a `mir::Const`... well that seems a bit pointless as evaluating it is literally the only thing you can do with a `mir::Const`. ;)

Reply:
I don't really know what to say tbh ^^ I think I've laid out pretty clearly why its *not* an option to keep all of these `eval_x` methods around on `ty::Const` since they will behave incorrectly ^^

I suppose we could try make `mir::Const` have a similar API to `ty::Const` but that's not going to happen by keeping `eval_x` around on `ty::Const` it would have to happen by making `mir::Const` mirror `ty::Const`'s *new* API surface

Reply:
The notion of "evaluating a const" needs to exist. You can't be suggesting to remove that concept. So what you are doing is moving it somewhere else. I don't know what goes wrong with the existing methods that can't be fixed, but `eval_valtree` still exists in this PR and it's IMO impossible to remove. So I don't know what you are suggesting.

`mir::Const` has a variant that embeds `ty::Const`, and MIR consts need to be evaluated in some key places in codegen and the interpreter (where the latter can be in generic code). What I am saying is to take whatever invocation needs to occur there, and put that into `mir::Const::eval`/`ty::Const::eval`. Not sure what else to say... this operation must be possible, so this is just about where that code lives?

Reply:
I have no clue what "we would have to return the nested goals of the normalization to the caller" means, I would expect when I ask for a const to be evaluated I can invoke normalization in a way that doesn't generate any "nested goals" (but just says "irreducible" in that case).

Reply:
> it would have to happen by making mir::Const mirror ty::Const's new API surface

It might be good to lay down what that API surface is and discuss it before doing all the work of implementing it, since there's multiple stakeholders on these types whose input should be considered.

But an `eval` method IMO must be part of it. *Something* has to happen [here](https://github.com/rust-lang/rust/blob/cd3da000c0ea2e1ce77a18587a7acdb01bf7fd23/compiler/rustc_codegen_ssa/src/mir/constant.rs#L23-L29) and [here](https://github.com/rust-lang/rust/blob/c682aa162b0d41e21cc6748f4fecfe01efb69d1f/compiler/rustc_const_eval/src/interpret/eval_context.rs#L568-L596), including for the case where this `mir::Const` is actually a `ty::Const`. If you remove `eval` from `mir::Const` you'll have to put some other code here. What I'm saying is that's the code that should be in `mir::Const::eval`. I don't understand why that should be a problem...

Reply:
All these eval-wrappers btw that this PR removes, I am happy to see disappear. I didn't know we could make sure things are always sufficiently normalized that they become unnecessary.

Reply:
> The notion of "evaluating a const" needs to exist. You can't be suggesting to remove that concept. So what you are doing is moving it somewhere else.

Indeed I am not suggesting we remove the ability to turn an unevaluated constant into a valtree... You can do this through normalization which *already exists in the compiler* and *already has logic to evaluate fully concrete constants in it*.

**The only thing that is changing is that you must go through the type system normalization functions instead of the `eval_x` functions**. I do not know how I can make this clearer but I feel like I am constantly repeating myself.

As you say, the logic has to go somewhere which is why we're not removing the ability to turn an unevaluated constant into a valtree. All that is changing is that instead of calling `ty::Const::eval_x` you now call the normalization functions that exist for normalizing things in the type system. It is too error prone to keep `ty::Const::eval_x` otherwise due to what i said about normalization returning goals that must be proven.

You mentioned something has to happen ["here"](https://github.com/rust-lang/rust/blob/c682aa162b0d41e21cc6748f4fecfe01efb69d1f/compiler/rustc_const_eval/src/interpret/eval_context.rs#L568-L596). We *already* invoke normalization there. `monomorphize` calls `instantiate_mir_and_normalize_erasing_regions` which normalizes `ty::Const`.

> I have no clue what "we would have to return the nested goals of the normalization to the caller" means. I would expect when I ask for a const to be evaluated I can invoke normalization in a way that doesn't generate any "nested goals" (but just says "irreducible" in that case).

Evaluating a fully concrete constant can be done without any kind of "returning of goals" yes. Handling of arbitrary const arguments such as `Foo<<T as Trait>::ASSOC>` is more complicated in the same way that it is for types. The way we do normalization of types is done for *good reason* because having a `Ty::normalize` function would be error prone which is *why we are removing it from `ty::Const`*.

Reply:
To put it incredibly directly: this refactoring does not remove any capabilities from `ty::Const` you can still do *everything* you could before.

Reply:
>  All that is changing is that instead of calling ty::Const::eval_x you now call the normalization functions that exist for normalizing things in the type system. It is too error prone to keep ty::Const::eval_x otherwise due to what i said about normalization returning goals that must be proven.

But `monomorphize` doesn't return any goals either...? This still doesn't make sense to me I am afraid. :(

> The way we do normalization of types is done for good reason because having a Ty::normalize function would be error prone which is why we are removing it from ty::Const.

Having `tcx.normalize(ty, param_env)` *cannot possibly be more error-prone* than `ty.normalize(tcx, param_env)`. That's literally just shifting the arguments around...

Reply:
Okay I think I [got it in the other thread](https://github.com/rust-lang/rust/pull/130950#discussion_r1779737581). Thanks for your patience!

Now I am wondering how to best document this. But that only really becomes relevant when `eval_valtree` gets removed.

Reply:
Sorry for taking a month to get back. I think in general our docs on type system normalization are very subpar which is very linked to what's going on here. I'll try and write something up for *that* which should then amke docs for this significantly easier to write.